### ISSUE TITLE
tr does not support character classes, repeats, etc

### ISSUE THREAD

**Issue by kwantam** [uutils/coreutils#556]
This isn't super difficult to do, but keeping it allocation-free a la the `ExpandSet` iterator will be entertaining :)

**Comment by Shnatsel** [comment]
I have implemented an escape parser for tr, complete with unit tests: https://github.com/Shnatsel/rust-tr/blob/master/src/escape_parser.rs

It supports all GNU escape characters as well as octal sequences. Behavior is mirrored from GNU tr. It's not allocation-free, but since it's only run once I don't see why would that be a goal.

I'll try to adapt it for uutils tr and contribute it if the time permits.

**Comment by hbina** [comment]
I tried to do it allocation free here https://github.com/uutils/coreutils/pull/2502
But I also want to keep the code as sane as possible. Please help me review and improve the code!

**Comment by jtracey** [comment]
Did #2502 fix this?

**Comment by kwantam** [comment]
Not clear it's 100% there, but it certainly gets closer. This bug is old enough that it might make sense to close in any case. Once #2502 lands it would be good to open a new issue to track anything that remains to do (if indeed there is anything left to do).

**Comment by jtracey** [comment]
I believe it did land, hence the bump. :wink: 

@hbina can you chime in with what, if anything, you think is left? It sounds like the GNU tests are basically passing, modulo some CI environment issues.

**Comment by kwantam** [comment]
> I believe it did land, hence the bump

D'oh! :)

**Comment by tertsdiepraam** [comment]
Yep, if I'm not mistaken it's all in there! There are tests for all character classes (I think), ranges and repeats. So I think this can be closed.

**Comment by hbina** [comment]
IIRC it passes all the GNU tests so it should all be there
